assets/js/modules/adsense/setup/setup-main.js

```diff
@@ -0,0 +1,293 @@
+import {
+	AdBlockerWarning,
+} from '../common';
+const { useSelect, useDispatch } = Data;
+
+export default function SetupMain( { finishSetup } ) {
+	// Get settings.
+	const siteURL = useSelect( ( select ) => select( siteStoreName ).getReferenceSiteURL() );
+	const previousAccountID = useSelect( ( select ) => select( STORE_NAME ).getAccountID() );
+	const previousClientID = useSelect( ( select ) => select( STORE_NAME ).getClientID() );
+	const previousAccountStatus = useSelect( ( select ) => select( STORE_NAME ).getAccountStatus() );
+	const previousSiteStatus = useSelect( ( select ) => select( STORE_NAME ).getSiteStatus() );
+	const accountSetupComplete = useSelect( ( select ) => select( STORE_NAME ).getAccountSetupComplete() );
+	const siteSetupComplete = useSelect( ( select ) => select( STORE_NAME ).getSiteSetupComplete() );
+
+	// Check whether a change submission is in progress.
+	const isDoingSubmitChanges = useSelect( ( select ) => select( STORE_NAME ).isDoingSubmitChanges() );
+
+	// Check whether settings differ from server and are valid.
+	const canSubmitChanges = useSelect( ( select ) => select( STORE_NAME ).canSubmitChanges() );
+
+	// Determine account.
+	const accounts = useSelect( ( select ) => select( STORE_NAME ).getAccounts() );
+	const accountID = determineAccountID( {
+		accounts,
+		previousAccountID,
+	} );
+
+	// Determine client.
+	const clients = useSelect( ( select ) => select( STORE_NAME ).getClients( accountID ) );
+	const clientID = determineClientID( {
+		clients,
+		previousClientID,
+	} );
+
+	// Get additional information to determine account and site status.
+	const alerts = useSelect( ( select ) => select( STORE_NAME ).getAlerts( accountID ) );
+	const urlChannels = useSelect( ( select ) => select( STORE_NAME ).getURLChannels( clientID ) );
+	const error = useSelect( ( select ) => select( STORE_NAME ).getError() );
+
+	// Determine account and site status.
+	const accountStatus = determineAccountStatus( {
+		accounts,
+		clients,
+		alerts,
+		error,
+		previousAccountID,
+		previousClientID,
+	} );
+	const siteStatus = determineSiteStatus( {
+		urlChannels,
+		siteURL,
+	} );
+
+	const {
+		setAccountID,
+		setClientID,
+		setAccountStatus,
+		setSiteStatus,
+		setAccountSetupComplete,
+		setSiteSetupComplete,
+		setUseSnippet,
+		submitChanges,
+	} = useDispatch( STORE_NAME );
+
+	// Update current account ID setting on-the-fly.
+	useEffect( () => {
+		// Don't do anything if setting has not loaded yet or if account ID cannot be determined.
+		if ( 'undefined' === typeof previousAccountID || 'undefined' === typeof accountID ) {
+			return;
+		}
+		setAccountID( accountID );
+	}, [ previousAccountID, accountID ] );
+
+	// Update current client ID setting on-the-fly.
+	useEffect( () => {
+		// Don't do anything if setting has not loaded yet or if client ID cannot be determined.
+		if ( 'undefined' === typeof previousClientID || 'undefined' === typeof clientID ) {
+			return;
+		}
+		setClientID( clientID );
```

Comment:
This effect is called every time `clientID` is updated, but then it updates `clientID`. That shouldn't cause infinite re-renders as `clientID` won't show as updated by the store so won't run this _again_, but it seems like this is run one-too-many-times already. Why `setClientID` after it's changed?

Reply:
The `clientID` value that is set here does not come from the setting, but from the client detection logic (`determineClientID`). As soon as that detection logic finds the client ID, it needs to be stored in the setting. The same applies to account ID, account status, and site status, those all have settings, but they're actually automatically determined by logic.